pulsar-common/src/main/java/org/apache/pulsar/common/util/SslContextRefresher.java

```diff
@@ -0,0 +1,146 @@
+        this.watchKeys = Sets.newHashSet();
+        this.files = Sets.newHashSet();
+        this.delay = delay;
+        this.timeunit = timeunit;
+        this.eventLoopGroup = eventLoopGroup;
+
+        files.add(trustCertsFilePath);
+        files.add(certificateFilePath);
+        files.add(keyFilePath);
+
+        // Remove all Null Or Empty Files
+        files.removeIf(file -> Strings.isNullOrEmpty(file));
+
+        this.sslContext = SecurityUtility.createNettySslContextForServer(tlsAllowInsecureConnection,
+                tlsTrustCertsFilePath, tlsCertificateFilePath, tlsKeyFilePath, tlsCiphers, tlsProtocols,
+                tlsRequireTrustedClientCertOnConnect);
+
+        for (String file : files) {
+            LOG.info("File {}", file);
+            Path p = Paths.get(file).getParent();
+            WatchService watchService = p.getFileSystem().newWatchService();
+            WatchKey key = p.register(watchService, StandardWatchEventKinds.ENTRY_CREATE,
+                    StandardWatchEventKinds.ENTRY_DELETE, StandardWatchEventKinds.ENTRY_MODIFY,
+                    StandardWatchEventKinds.OVERFLOW);
+            watchKeys.add(key);
+        }
+
+        if (LOG.isDebugEnabled()) {
+            LOG.debug("Certs will be refreshed every {} minutes", delay);
+        }
+
+        if (delay > 0) {
+            run();
+        }
+    }
+
+    public void run() {
+        for (WatchKey key : watchKeys) {
+            List<WatchEvent<?>> events = key.pollEvents();
+            for (WatchEvent<?> e : events) {
+                String file = ((Path) key.watchable()).toFile().getAbsolutePath();
+                if (e.kind().equals(StandardWatchEventKinds.OVERFLOW)) {
+                    LOG.warn("Overflow occured for file {} - rebuilding sslContext", file);
+                    buildSSLContext();
+                    break;
+                } else if (files.contains(file)) {
+                    LOG.info("Changed found for file {} - rebuilding sslContext", file);
+                    buildSSLContext();
+                    break;
+                } else if (LOG.isDebugEnabled()) {
+                    LOG.debug("Ignoring changes for file {}.", file);
+                }
+            }
+        }
+        this.eventLoopGroup.schedule(this, this.delay, this.timeunit);
+    }
+
+    public void buildSSLContext() {
+        SslContext sslContext;
+        try {
+            sslContext = SecurityUtility.createNettySslContextForServer(tlsAllowInsecureConnection,
+                    tlsTrustCertsFilePath, tlsCertificateFilePath, tlsKeyFilePath, tlsCiphers, tlsProtocols,
+                    tlsRequireTrustedClientCertOnConnect);
+        } catch (GeneralSecurityException | IOException e) {
+            LOG.error("Error occured while trying to create sslContext - using previous one.");
+            return;
+        }
+        this.sslContext = sslContext;
+    }
+
+    public SslContext get() {
```

Comment:
Instead of using a background thread, couldn't we just check a timestamp here to check whether the cached context is still valid?

Reply:
We could even just check the stat of the file each time, without having to configure a caching time. If the file hasn't changed since last time, then reuse context.

Reply:
Agreed - will make the changes accordingly.